bionic/core/task_execution.py

```diff
@@ -751,13 +752,35 @@ def _local_artifact_from_value(self, value):
                 )
             ) from e
 
-        value_hash = protocol.tokenize_file(value_path)
+        value_hash = self._generate_value_hash(value, value_path, context)
         return Artifact(
             url=url_from_path(value_path),
             content_hash=value_hash,
         )
         return value_path
 
+    def _generate_value_hash(self, value, value_path, context):
+        protocol = self.desc_metadata.protocol
+        code_versioning_policy = self.func_attrs.code_versioning_policy
+        code_version = code_versioning_policy.version
+        versioning_policy = context.core.versioning_policy
+        try:
+            bytecode_hash = CodeHasher.hash(
+                type(value),
+                code_versioning_policy.suppress_bytecode_warnings
+                or versioning_policy.ignore_bytecode_exceptions
+                or not code_version.includes_bytecode,
+            )
+        except Exception:
+            bytecode_hash = None
+            if (
+                not versioning_policy.ignore_bytecode_exceptions
+                and code_version.includes_bytecode
+            ):
+                raise
+
+        return protocol.tokenize_file(value_path) + protocol.tokenize(bytecode_hash)
```

Comment:
`protocol.tokenize` is really only intended to be applied to valid values for that protocol. Since `bytecode_hash` is already a hash, could we just append it directly to the tokenized file? (Or make this hashing logic the responsibility of the protocol?)

Reply:
Taking your [previous comment](https://github.com/square/bionic/pull/319#discussion_r549390370) into consideration, I think it makes sense to make this the responsibility of the protocol.

Reply:
I moved the logic into the protocol and appended it to the tokenized file.